Scanning of char into string possible in TextFormat but not in BinaryFormat

**Describe the bug**
First of I think this is a bug but im not sure.  
If you scan a 'char' into a go string it works if the wire? format is text and not binary.
As soon as it is binary it does not work anymore.


**To Reproduce**
Steps to reproduce the behavior:

```go
package main

import (
	"context"
	"os"

	"github.com/jackc/pgx/v5"
)

func main() {

	ctx := context.Background()

	conn, err := pgx.Connect(ctx, os.Getenv("DATABASE_URL"))
	if err != nil {
		panic(err)
	}
	defer conn.Close(ctx)

	rows, err := conn.Query(ctx, `SELECT confupdtype FROM pg_constraint LIMIT 1;`)
	if err != nil {
		panic(err)
	}
	defer rows.Close()

	for rows.Next() {
		updateType := ""
		if err := rows.Scan(&updateType); err != nil {
			panic(err)
		}
	}
}
```

go.mod

```
module bugreport

go 1.22.2

require github.com/jackc/pgx/v5 v5.7.4

require (
	github.com/jackc/pgpassfile v1.0.0 // indirect
	github.com/jackc/pgservicefile v0.0.0-20240606120523-5a60cdf6a761 // indirect
	golang.org/x/crypto v0.31.0 // indirect
	golang.org/x/text v0.21.0 // indirect
)
```

Please run your example with the race detector enabled. For example, `go run -race main.go` or `go test -race`.

**Expected behavior**
I expect the `rows.Scan` to work regardless of the format. Especially since I as the user of the library don't choose the format.


**Actual behavior**
```
panic: can't scan into dest[0] (col: confupdtype): cannot scan char (OID 18) in binary format into *string

goroutine 1 [running]:
main.main()
        .../pgx_bug/main.go:29 +0x30a
exit status 2
```

**Version**
 - Go: `go version go1.22.2 windows/amd64`
 - PostgreSQL: `PostgreSQL 17.2 on x86_64-windows, compiled by msvc-19.41.34123, 64-bit`
 - pgx: `v5.7.4`

**Additional context**
Add any other context about the problem here.

**Workaround**

We just fixed our code to use `rune` instead of `string` which is probably better anyway but I thought i'd share the problem.

--- a/pgtype/qchar.go
+++ b/pgtype/qchar.go
@@ -64,6 +64,8 @@ func (QCharCodec) PlanScan(m *Map, oid uint32, format int16, target any) ScanPla
 			return scanPlanQcharCodecByte{}
 		case *rune:
 			return scanPlanQcharCodecRune{}
+		case *string:
+			return scanPlanQcharCodecString{}
 		}
 	}
 
@@ -110,6 +112,26 @@ func (scanPlanQcharCodecRune) Scan(src []byte, dst any) error {
 	} else {
 		*r = rune(src[0])
 	}
+
+	return nil
+}
+
+type scanPlanQcharCodecString struct{}
+
+func (scanPlanQcharCodecString) Scan(src []byte, dst any) error {
+	if src == nil {
+		return fmt.Errorf("cannot scan NULL into %T", dst)
+	}
+
+	if len(src) > 1 {
+		return fmt.Errorf(`invalid length for "char": %v`, len(src))
+	}
+
+	p := dst.(*string)
+	// Copy the raw byte so the result matches the text-format *string scan path
+	// (string(src)) byte-for-byte. string(src[0]) would instead UTF-8-encode the
+	// byte as a code point, diverging for byte values >= 128.
+	*p = string(src)
 
 	return nil
 }

--- a/pgtype/qchar_test.go
+++ b/pgtype/qchar_test.go
@@ -5,6 +5,7 @@ import (
 	"math"
 	"testing"
 
+	"github.com/jackc/pgx/v5/pgtype"
 	"github.com/jackc/pgx/v5/pgxtest"
 )
 
@@ -21,4 +22,72 @@ func TestQcharTranscode(t *testing.T) {
 
 	// Can only test with known OIDs as rune and byte would be considered numbers.
 	pgxtest.RunValueRoundTripTests(context.Background(), t, defaultConnTestRunner, pgxtest.KnownOIDQueryExecModes, `"char"`, tests)
+}
+
+// TestQcharCodecPlanScanString is a regression test for https://github.com/jackc/pgx/issues/2314.
+//
+// Scanning a "char" column (OID 18) into *string used to succeed in TextFormat but
+// fail in BinaryFormat with "cannot scan char (OID 18) in binary format into *string".
+// Both formats must now produce the same result.
+func TestQcharCodecPlanScanString(t *testing.T) {
+	m := pgtype.NewMap()
+
+	for _, tt := range []struct {
+		name   string
+		format int16
+	}{
+		{"text", pgtype.TextFormatCode},
+		{"binary", pgtype.BinaryFormatCode},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			var s string
+			plan := m.PlanScan(pgtype.QCharOID, tt.format, &s)
+			if plan == nil {
+				t.Fatalf("PlanScan returned nil plan for *string in %s format", tt.name)
+			}
+			if err := plan.Scan([]byte{'a'}, &s); err != nil {
+				t.Fatalf("Scan failed in %s format: %v", tt.name, err)
+			}
+			if s != "a" {
+				t.Fatalf("Scan result mismatch in %s format: got %q want %q", tt.name, s, "a")
+			}
+		})
+	}
+
+	// Empty src must produce empty string (mirrors *byte / *rune zero-value behavior).
+	t.Run("empty-binary", func(t *testing.T) {
+		var s string = "x"
+		plan := m.PlanScan(pgtype.QCharOID, pgtype.BinaryFormatCode, &s)
+		if err := plan.Scan([]byte{}, &s); err != nil {
+			t.Fatalf("Scan failed: %v", err)
+		}
+		if s != "" {
+			t.Fatalf("empty src: got %q want %q", s, "")
+		}
+	})
+
+	// 0xC8 (200): a byte >= 128. Both formats must yield the raw 1-byte string
+	// "\xc8", not the 2-byte UTF-8 encoding "\xc3\x88". This is the case that
+	// catches string(src[0]) UTF-8-encoding the byte instead of copying it.
+	t.Run("non-ascii-byte", func(t *testing.T) {
+		for _, format := range []int16{pgtype.TextFormatCode, pgtype.BinaryFormatCode} {
+			var s string
+			plan := m.PlanScan(pgtype.QCharOID, format, &s)
+			if err := plan.Scan([]byte{0xC8}, &s); err != nil {
+				t.Fatalf("format %d: scan failed: %v", format, err)
+			}
+			if s != "\xc8" {
+				t.Fatalf("format %d: got %q (% x) want %q", format, s, s, "\xc8")
+			}
+		}
+	})
+
+	// Multi-byte src is an invalid "char" payload and must error.
+	t.Run("too-long", func(t *testing.T) {
+		var s string
+		plan := m.PlanScan(pgtype.QCharOID, pgtype.BinaryFormatCode, &s)
+		if err := plan.Scan([]byte("ab"), &s); err == nil {
+			t.Fatalf("expected error for 2-byte src, got nil (s=%q)", s)
+		}
+	})
 }

